I've added a unit test. Without my patch, we get the following error due to an error reading a corrupted state store.

> Caused by: java.io.IOException: Cannot read /hdfs-federation/MembershipState/randomString--459989902-randomString--826693283-randomString--606850105 for record MembershipState
	at org.apache.hadoop.hdfs.server.federation.store.driver.impl.StateStoreFileBaseImpl.getRecord(StateStoreFileBaseImpl.java:290)
	at org.apache.hadoop.hdfs.server.federation.store.driver.impl.StateStoreFileBaseImpl.get(StateStoreFileBaseImpl.java:222)

--- a/hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/store/driver/impl/StateStoreFileBaseImpl.java
+++ b/hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/store/driver/impl/StateStoreFileBaseImpl.java
@@ -349,25 +349,15 @@ public abstract class StateStoreFileBaseImpl
     for (Entry<String, T> entry : toWrite.entrySet()) {
       String recordPath = entry.getKey();
       String recordPathTemp = recordPath + "." + now() + TMP_MARK;
-      BufferedWriter writer = getWriter(recordPathTemp);
-      boolean recordWrittenSuccessfully = false;
-      try {
+      boolean recordWrittenSuccessfully = true;
+      try (BufferedWriter writer = getWriter(recordPathTemp)) {
         T record = entry.getValue();
         String line = serializeString(record);
         writer.write(line);
-        recordWrittenSuccessfully = true;
       } catch (IOException e) {
         LOG.error("Cannot write {}", recordPathTemp, e);
+        recordWrittenSuccessfully = false;
         success = false;
-      } finally {
-        if (writer != null) {
-          try {
-            writer.close();
-          } catch (IOException e) {
-            recordWrittenSuccessfully = false;
-            LOG.error("Cannot close the writer for {}", recordPathTemp, e);
-          }
-        }
       }
       // Commit
       if (recordWrittenSuccessfully && !rename(recordPathTemp, recordPath)) {

--- a/hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/store/driver/impl/StateStoreFileBaseImpl.java
+++ b/hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/store/driver/impl/StateStoreFileBaseImpl.java
@@ -85,17 +85,9 @@ public abstract class StateStoreFileBaseImpl
    * @param path Path of the record to write.
    * @return Writer for the record.
    */
-  protected abstract <T extends BaseRecord> BufferedWriter getWriter(
-      String path);
-
-  /**
-   * Convenience method to allow to mocking of protected
-   * {@link #getWriter(String)}
-   */
   @VisibleForTesting
-  public BufferedWriter getBufferedWriter(String path) {
-    return getWriter(path);
-  }
+  public abstract <T extends BaseRecord> BufferedWriter getWriter(
+      String path);
 
   /**
    * Check if a path exists.
@@ -357,16 +349,16 @@ public abstract class StateStoreFileBaseImpl
     for (Entry<String, T> entry : toWrite.entrySet()) {
       String recordPath = entry.getKey();
       String recordPathTemp = recordPath + "." + now() + TMP_MARK;
-      BufferedWriter writer = getBufferedWriter(recordPathTemp);
-      boolean recordWrittenSuccessfully = true;
+      BufferedWriter writer = getWriter(recordPathTemp);
+      boolean recordWrittenSuccessfully = false;
       try {
         T record = entry.getValue();
         String line = serializeString(record);
         writer.write(line);
+        recordWrittenSuccessfully = true;
       } catch (IOException e) {
         LOG.error("Cannot write {}", recordPathTemp, e);
         success = false;
-        recordWrittenSuccessfully = false;
       } finally {
         if (writer != null) {
           try {

--- a/hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/store/driver/impl/StateStoreFileBaseImpl.java
+++ b/hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/store/driver/impl/StateStoreFileBaseImpl.java
@@ -358,6 +358,7 @@ public abstract class StateStoreFileBaseImpl
       String recordPath = entry.getKey();
       String recordPathTemp = recordPath + "." + now() + TMP_MARK;
       BufferedWriter writer = getBufferedWriter(recordPathTemp);
+      boolean recordWrittenSuccessfully = true;
       try {
         T record = entry.getValue();
         String line = serializeString(record);
@@ -365,17 +366,19 @@ public abstract class StateStoreFileBaseImpl
       } catch (IOException e) {
         LOG.error("Cannot write {}", recordPathTemp, e);
         success = false;
+        recordWrittenSuccessfully = false;
       } finally {
         if (writer != null) {
           try {
             writer.close();
           } catch (IOException e) {
+            recordWrittenSuccessfully = false;
             LOG.error("Cannot close the writer for {}", recordPathTemp, e);
           }
         }
       }
       // Commit
-      if (success && !rename(recordPathTemp, recordPath)) {
+      if (recordWrittenSuccessfully && !rename(recordPathTemp, recordPath)) {
         LOG.error("Failed committing record into {}", recordPath);
         success = false;
       }

--- a/hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/store/driver/impl/StateStoreFileImpl.java
+++ b/hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/store/driver/impl/StateStoreFileImpl.java
@@ -31,6 +31,7 @@ import java.util.Collections;
 import java.util.List;
 
 import org.apache.commons.lang3.ArrayUtils;
+import org.apache.hadoop.classification.VisibleForTesting;
 import org.apache.hadoop.hdfs.server.federation.router.RBFConfigKeys;
 import org.apache.hadoop.hdfs.server.federation.store.records.BaseRecord;
 import org.slf4j.Logger;
@@ -125,7 +126,8 @@ public class StateStoreFileImpl extends StateStoreFileBaseImpl {
   }
 
   @Override
-  protected <T extends BaseRecord> BufferedWriter getWriter(String filename) {
+  @VisibleForTesting
+  public <T extends BaseRecord> BufferedWriter getWriter(String filename) {
     BufferedWriter writer = null;
     try {
       LOG.debug("Writing file: {}", filename);

--- a/hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/store/driver/impl/StateStoreFileSystemImpl.java
+++ b/hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/store/driver/impl/StateStoreFileSystemImpl.java
@@ -28,6 +28,7 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 
+import org.apache.hadoop.classification.VisibleForTesting;
 import org.apache.hadoop.fs.FSDataInputStream;
 import org.apache.hadoop.fs.FSDataOutputStream;
 import org.apache.hadoop.fs.FileStatus;
@@ -139,7 +140,8 @@ public class StateStoreFileSystemImpl extends StateStoreFileBaseImpl {
   }
 
   @Override
-  protected <T extends BaseRecord> BufferedWriter getWriter(String pathName) {
+  @VisibleForTesting
+  public <T extends BaseRecord> BufferedWriter getWriter(String pathName) {
     BufferedWriter writer = null;
     Path path = new Path(pathName);
     try {

--- a/hadoop-hdfs-project/hadoop-hdfs-rbf/src/test/java/org/apache/hadoop/hdfs/server/federation/store/driver/TestStateStoreFileSystem.java
+++ b/hadoop-hdfs-project/hadoop-hdfs-rbf/src/test/java/org/apache/hadoop/hdfs/server/federation/store/driver/TestStateStoreFileSystem.java
@@ -111,7 +111,7 @@ public class TestStateStoreFileSystem extends TestStateStoreDriverBase {
       BufferedWriter spyWriter = spy(writer);
       doThrow(IOException.class).when(spyWriter).write(any(String.class));
       return spyWriter;
-    }).when(driver).getBufferedWriter(any());
+    }).when(driver).getWriter(any());
 
     testInsertWithErrorDuringWrite(driver, MembershipState.class);
   }

